platform/solr/solr-factory/src/main/java/org/codice/solr/client/solrj/SolrClient.java

```diff
@@ -0,0 +1,1030 @@
+   * to become available. This method is different from an active ping as it might be basing its
+   * answer on cached information. It might, however, initiate a background ping.
+   *
+   * <p><i>Note:</i> Nothing will happen and the method will return right away if the client was
+   * already closed.
+   *
+   * <p><i>Note:</i> Returning <code>true</code> doesn't guarantee the next call to any methods will
+   * succeed as the server or the core might become unavailable at any time without notice.
+   *
+   * @param timeout the maximum time to wait
+   * @param unit the time unit of the timeout argument
+   * @return <code>false</code> if the Solr server this client is connected or the corresponding
+   *     core is not available; <code>true</code> otherwise
+   * @throws InterruptedException if the current thread was interrupted while waiting
+   */
+  public boolean isAvailable(long timeout, TimeUnit unit) throws InterruptedException;
+
+  /**
+   * This method is the same as {@link #isAvailable()} but allows a caller to provide a listener
+   * that will be called back with a boolean flag representing the new availability for this Solr
+   * client whenever it changes in the future. The listener will also be called back right away with
+   * the current availability of the client. This method is different from an active ping as it
+   * might be basing its initial answer on cached information. It might, however, initiate a
+   * background ping.
+   *
+   * <p><i>Note:</i> Nothing will happen and the method will return right away if the client was
+   * already closed.
+   *
+   * <p><i>Note:</i> Returning <code>true</code> doesn't guarantee the next call to any methods will
+   * succeed as the server or the core might become unavailable at any time without notice.
+   *
+   * @param listener a listener to notify whenever the Solr client's availability changes
+   * @return <code>false</code> if the Solr server this client is connected or the corresponding
+   *     core is not available; <code>true</code> otherwise
+   */
+  public boolean isAvailable(Listener listener);
+
+  /**
+   * Registers an initializer to be called when this Solr client finishes initializing and becomes
+   * available. The initializer is called right away if this client is already available. The
+   * initializer will never be called twice.
+   *
+   * <p><i>Note:</i> Nothing will happen and the method will return right away if the client was
+   * already closed.
+   *
+   * <p><i>Note:</i> The client might no longer be available by the time the initializer is called
+   * as the server or the core might become unavailable at any time without notice. It is, however,
+   * guaranteed that the client was available at least once.
+   *
+   * @param initializer an initializer to notify when the Solr client first become available
+   */
+  public void whenAvailable(Initializer initializer);
```

Comment:
❓Could we simplify the interface a little and combine this method with the `isAvailable(Listener)`? Not sure we need both.

Reply:
Actually, this is much simpler from a user's perspective. You do not have to worry about removing your listener and dealing with being called twice. A listener is also called for both state changes (i.e. available and not available) as this one only gets called for the state becoming available. If you look at the banana code, you will see it makes things much easier.

Plus, to be in sync with the catalog provider, I didn't want to provide a remove listener mechanism as the typically listeners will want to be there always wheres as initializers do not.